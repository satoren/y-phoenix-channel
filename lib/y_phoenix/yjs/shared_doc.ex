defmodule YPhoenix.SharedDoc do
  @moduledoc """
  """
  use GenServer

  require Logger
  alias Yex.Sync

  @idle_timeout 15000
  defmodule Persistence do
    @moduledoc """
    Persistence behavior for SharedDoc
    """

    @callback bind(doc_name :: String.t(), doc :: Yex.Doc.t()) :: term()
    @callback unbind(term :: term(), doc_name :: String.t(), doc :: Yex.Doc.t()) :: :ok
    @callback update_v1(
                term :: term(),
                update :: binary(),
                doc_name :: String.t(),
                doc :: Yex.Doc.t()
              ) :: term()

    @optional_callbacks update_v1: 4, unbind: 3
  end

  defmodule LocalPubSub do
    @moduledoc """
    LocalPubSub behavior for SharedDoc
    Used to notify SharedDoc users of updates.
    """

    @callback monitor_count(doc_name :: String.t()) :: integer()
    @callback broadcast(doc_name :: String.t(), message :: term(), exclude_origin :: term()) ::
                :ok

    @callback monitor(doc_name :: String.t()) :: :ok
    @callback demonitor(doc_name :: String.t()) :: :ok
  end

  def send_yjs_message(server, message) when is_binary(message) do
    send(GenServer.whereis(server), {:yjs, message, self()})
  end

  def start_sync(server, step1_message) do
    send(GenServer.whereis(server), {:start_sync, step1_message, self()})
  end

  def doc_name(server) do
    GenServer.call(server, :doc_name)
  end

  def start_link(param, option \\ []) do
    GenServer.start_link(__MODULE__, param, option)
  end

  def start(param, option \\ []) do
    GenServer.start(__MODULE__, param, option)
  end

  @impl true
  def init(option) do
    doc_name = Keyword.fetch!(option, :doc_name)
    persistence = Keyword.get(option, :persistence)
    timeout = Keyword.get(option, :idle_timeout, @idle_timeout)
    pg_scope = Keyword.get(option, :pg_scope, nil)
    local_pubsub = Keyword.get(option, :local_pubsub, nil)
    doc = Yex.Doc.new()
    {:ok, awareness} = Yex.Awareness.new(doc)

    Yex.Awareness.clean_local_state(awareness)

    persistence_state =
      if function_exported?(persistence, :bind, 2) do
        persistence.bind(doc_name, doc)
      end

    {:ok, step1_data} = Sync.get_sync_step1(doc)
    message = Yex.Sync.message_encode!({:sync, step1_data})
    step1 = {:yjs, message, self()}

    if local_pubsub != nil do
      local_pubsub.broadcast(
        doc_name,
        step1,
        ""
      )
    end

    if pg_scope != nil do
      :pg.join(pg_scope, doc_name, self())
      {_group_monitor_ref, pids} = :pg.monitor(pg_scope, doc_name)

      pids
      |> Enum.reject(&(&1 == self()))
      |> Enum.each(fn pid ->
        send(pid, step1)
      end)
    end

    Yex.Doc.monitor_update_v1(doc)
    Yex.Awareness.monitor_change(awareness)

    {:ok,
     %{
       doc: doc,
       awareness: awareness,
       doc_name: doc_name,
       persistence: persistence,
       persistence_state: persistence_state,
       timeout: timeout,
       pg_scope: pg_scope,
       local_pubsub: local_pubsub
     }, timeout}
  end

  @impl true
  def handle_call(:doc_name, _from, state) do
    {:reply, state.doc_name, state}
  end

  @impl true
  def handle_info({:start_sync, message, from}, state) when is_binary(message) do
    with {:ok, {:sync, sync_message}} <- Yex.Sync.message_decode(message),
         {:ok, reply} <- Sync.read_sync_message(sync_message, state.doc, "#{inspect(from)}"),
         {:ok, sync_message} = Yex.Sync.message_encode({:sync, reply}) do
      send(from, {:yjs, sync_message, self()})

      with {:ok, step1} <- Sync.get_sync_step1(state.doc),
           {:ok, step1} <- Yex.Sync.message_encode({:sync, step1}) do
        send(from, {:yjs, step1, self()})
      else
        error ->
          error
      end

      awareness_clients = Yex.Awareness.get_client_ids(state.awareness)

      with true <- length(awareness_clients) > 0,
           {:ok, awareness_update} <-
             Yex.Awareness.encode_update(state.awareness, awareness_clients) do
        send(from, {:yjs, Yex.Sync.message_encode!({:awareness, awareness_update}), self()})
      else
        false -> :ok
        error -> error
      end
    else
      error ->
        error
    end

    {:noreply, state, state.timeout}
  end

  @impl true
  def handle_info({:yjs, message, from}, state) when is_binary(message) do
    case Yex.Sync.message_decode(message) do
      {:ok, message} ->
        handle_yjs_message(message, from, state)

      error ->
        Logger.error(error)
        {:noreply, state, state.timeout}
    end
  end

  @impl true
  def handle_info({:update_v1, update, origin, _doc}, state) do
    state =
      if function_exported?(state.persistence, :update_v1, 4) do
        persistence_state =
          state.persistence.update_v1(state.persistence_state, update, state.doc_name, state.doc)

        put_in(state, :persistence_state, persistence_state)
      else
        state
      end

    with {:ok, s} <- Sync.get_update(update),
         {:ok, message} = Yex.Sync.message_encode({:sync, s}) do
      broadcast_to_group_process(message, origin, state)

      broadcast_to_users(message, origin, state)
    else
      error ->
        error
    end

    {:noreply, state, state.timeout}
  end

  @impl true
  def handle_info(
        {:awareness_change, %{removed: removed, added: added, updated: updated}, origin,
         awareness},
        state
      ) do
    changed_clients = added ++ updated ++ removed

    with {:ok, update} <- Yex.Awareness.encode_update(awareness, changed_clients),
         {:ok, message} = Yex.Sync.message_encode({:awareness, update}) do
      broadcast_to_group_process(message, origin, state)
      broadcast_to_users(message, origin, state)
    else
      error ->
        Logger.error(error)
        error
    end

    {:noreply, state, state.timeout}
  end

  def handle_info({_ref, :join, _group, pids}, %{doc: doc} = state) do
    with {:ok, s} <- Sync.get_sync_step1(doc),
         {:ok, step1} <- Yex.Sync.message_encode({:sync, s}) do
      pids
      |> Enum.each(fn pid ->
        send(pid, {:yjs, step1, self()})
      end)
    else
      error ->
        Logger.error(error)
        error
    end

    {:noreply, state, state.timeout}
  end

  def handle_info({_ref, :leave, _group, _pids}, state) do
    {:noreply, state, state.timeout}
  end

  def handle_info(:timeout, state) do
    if should_exit?(state) do
      {:stop, :normal, state}
    else
      {:noreply, state, state.timeout}
    end
  end

  @impl true
  def terminate(_reason, state) do
    if function_exported?(state.persistence, :unbind, 3) do
      state.persistence.unbind(state.persistence_state, state.doc_name, state.doc)
    end

    :ok
  end

  defp handle_yjs_message({:sync, sync_message}, from, state) do
    with {:ok, reply} <- Sync.read_sync_message(sync_message, state.doc, "#{inspect(from)}"),
         {:ok, sync_message} = Yex.Sync.message_encode({:sync, reply}) do
      send(from, {:yjs, sync_message, self()})
    else
      error ->
        error
    end

    {:noreply, state, state.timeout}
  end
  defp handle_yjs_message({:awareness, message}, from, state) do
    Yex.Awareness.apply_update(state.awareness, message)
    {:noreply, state, state.timeout}
  end

  defp handle_yjs_message(_, from, state) do

    # unsupported_message
    {:noreply, state, state.timeout}
  end

  defp should_exit?(state) do
    state.local_pubsub && state.local_pubsub.monitor_count(state.doc_name) === 0
  end

  defp broadcast_to_users(message, origin, state) do
    if state.local_pubsub != nil do
      state.local_pubsub.broadcast(
        state.doc_name,
        {:yjs, message, self()},
        origin
      )
    end
  end

  defp broadcast_to_group_process(message, _origin, state) do
    if state.pg_scope != nil do
      :pg.get_members(state.pg_scope, state.doc_name)
      |> Enum.reject(&(&1 == self()))
      |> Enum.each(fn pid ->
        send(pid, {:yjs, message, self()})
      end)
    end
  end
end
