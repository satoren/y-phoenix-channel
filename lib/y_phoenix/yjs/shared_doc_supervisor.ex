defmodule YPhoenix.SharedDocSupervisor do
  @moduledoc """
  """

  alias YPhoenix.SharedDoc
  use Supervisor

  defmodule SharedDocLocalPubsub do
    @behaviour YPhoenix.SharedDoc.LocalPubSub

    def monitor_count(doc_name) do
      Registry.lookup(__MODULE__, doc_name) |> Enum.count()
    end

    def broadcast(doc_name, message, exclude_origin) do
      Registry.dispatch(__MODULE__, doc_name, fn entries ->
        entries
        |> Enum.reject(fn {_pid, origin} -> origin === exclude_origin end)
        |> Enum.each(fn {pid, _origin} -> send(pid, message) end)
      end)
    end

    def monitor(doc_name) do
      {:ok, _} = Registry.register(__MODULE__, doc_name, "#{inspect(self())}")
      :ok
    end

    def demonitor(doc_name) do
      :ok = Registry.unregister(__MODULE__, doc_name)
    end
  end

  @registry YPhoenix.SharedDocSupervisor.SharedDocRegistry
  @local_pubsub YPhoenix.SharedDocSupervisor.SharedDocLocalPubsub
  @dynamic_supervisor YPhoenix.SharedDocSupervisor.DynamicSupervisor

  def start_link(init_arg) do
    Supervisor.start_link(__MODULE__, init_arg, name: __MODULE__)
  end

  @impl true
  def init(init_arg) do
    pg_scope = Keyword.get(init_arg, :pg_scope, YPhoenix.SharedDocScope)

    children = [
      %{
        id: :pg,
        start: {:pg, :start_link, [pg_scope]}
      },
      {Registry, keys: :unique, name: @registry},
      {Registry, keys: :duplicate, name: @local_pubsub},
      {@dynamic_supervisor, [local_pubsub: @local_pubsub] ++ init_arg}
    ]

    Supervisor.init(children, strategy: :one_for_one)
  end

  def start_child(doc_name) do
    name = via_name(doc_name)

    @dynamic_supervisor.start_child(doc_name: doc_name, name: name)

    try do
      # check started
      SharedDoc.doc_name(name)
    catch
      _ ->
        @dynamic_supervisor.start_child(doc_name: doc_name, name: name)
    end

    {:ok, name}
  end

  def via_name(doc_name) do
    {:via, Registry, {@registry, doc_name}}
  end

  defdelegate monitor_shared_doc_update(doc_name), to: @local_pubsub, as: :monitor
  defdelegate demonitor_shared_doc_update(doc_name), to: @local_pubsub, as: :demonitor
  defdelegate monitor_count(doc_name), to: @local_pubsub

  defdelegate broadcast_shared_doc_update(doc_name, message, exclude_origin),
    to: @local_pubsub,
    as: :broadcast
end

defmodule YPhoenix.SharedDocSupervisor.DynamicSupervisor do
  use DynamicSupervisor

  defmodule ChildSpec do
    def start_link(extra_arguments, option) do
      name = Keyword.fetch!(option, :name)
      option = option |> Keyword.delete(:name)
      YPhoenix.SharedDoc.start_link(option ++ extra_arguments, name: name)
    end

    def child_spec(opts) do
      %{
        id: YPhoenix.SharedDoc,
        start: {__MODULE__, :start_link, [opts]}
      }
    end
  end

  def start_link(init_arg) do
    DynamicSupervisor.start_link(__MODULE__, init_arg, name: __MODULE__)
  end

  def start_child(args) do
    spec = {
      ChildSpec,
      args
    }

    DynamicSupervisor.start_child(__MODULE__, spec)
  end

  @impl true
  def init(init_arg) do
    DynamicSupervisor.init(
      strategy: :one_for_one,
      extra_arguments: [init_arg]
    )
  end
end
