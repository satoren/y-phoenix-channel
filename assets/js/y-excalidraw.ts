import type {
  BinaryFileData,
  Collaborator,
  ExcalidrawImperativeAPI,
} from "@excalidraw/excalidraw/types/types";
import * as Y from "yjs";
import type * as awarenessProtocol from "y-protocols/awareness";
import deepEqual from "fast-deep-equal";
import { NonDeletedExcalidrawElement } from "@excalidraw/excalidraw/types/element/types";

export class ExcalidrawBinding {
  subscriptions: (() => void)[] = [];
  collaborators: Map<string, Collaborator> = new Map();
  lastVersion: number | undefined = undefined;
  constructor(
    yarray: Y.Array<unknown>,
    api: ExcalidrawImperativeAPI,
    private awareness?: awarenessProtocol.Awareness,
  ) {
    this.subscriptions.push(
      api.onChange((elements) => {
        const version = getVersion(elements);
        if (this.lastVersion === version) {
          return;
        }
        this.lastVersion = version;

        const doc = yarray.doc;

        doc?.transact(() => {
          if (yarray.length < elements.length) {
            const add = Array.from(
              { length: elements.length - yarray.length },
              () => new Y.Map(),
            );
            yarray.push(add);
          } else if (yarray.length > elements.length) {
            yarray.delete(elements.length, yarray.length - elements.length);
          }

          for (let i = 0; i < elements.length; i++) {
            const map = yarray.get(i) as Y.Map<unknown>;
            const elem = elements[i];
            if (map.get("version") === elem.version) {
              continue;
            }

            for (const [key, value] of Object.entries(elem)) {
              const src = map.get(key);
              if (
                (typeof src === "object" && !deepEqual(map.get(key), value)) ||
                src !== value
              ) {
                map.set(key, value);
              }
            }
          }
        }, this);
      }),
    );

    yarray.observeDeep((_events, txn) => {
      if (txn.origin === this) {
        return;
      }

      const elements = [...api.getSceneElements()];

      for (const map of yarray) {
        if(!(map instanceof Y.Map)) {
          continue;
        }
        const id = map.get("id");
        const index = elements.findIndex((elem) => elem.id === id);
        if (index >= 0) {
          const version = map.get("version");
          if(version === elements[index].version) {
            elements[index] = map.toJSON() as NonDeletedExcalidrawElement;
          } else {
            elements.push(map.toJSON() as NonDeletedExcalidrawElement);
          }
        }
      }
    });

    api.updateScene({ elements: yarray.toJSON() });

    if (awareness) {
      this.subscriptions.push(
        api.onChange((_, state) => {
          awareness.setLocalStateField(
            "selectedElementIds",
            state.selectedElementIds,
          );
        }),
      );

      const awarenessChangeHandler = ({
        added,
        updated,
        removed,
      }: {
        added: number[];
        updated: number[];
        removed: number[];
      }) => {
        const states = awareness.getStates();

        const collaborators = new Map(this.collaborators);
        const update = [...added, ...updated];
        for (const id of update) {
          const state = states.get(id);
          if (!state) {
            continue;
          }

          collaborators.set(id.toString(), {
            pointer: state.pointer,
            button: state.button,
            selectedElementIds: state.selectedElementIds,
            username: state.user?.name,
            color: state.user?.color,
            avatarUrl: state.user?.avatarUrl,
            userState: state.user?.state,
          });
        }
        for (const id of removed) {
          collaborators.delete(id.toString());
        }
        collaborators.delete(awareness.clientID.toString());
        api.updateScene({
          collaborators,
        });
        this.collaborators = collaborators;
      };
      awareness.on("change", awarenessChangeHandler);
      this.subscriptions.push(() => {
        awareness.off("change", awarenessChangeHandler);
      });
    }
  }

  onPointerUpdate = (payload: {
    pointer: {
      x: number;
      y: number;
      tool: "pointer" | "laser";
    };
    button: "down" | "up";
  }) => {
    this.awareness?.setLocalStateField("pointer", payload.pointer);
    this.awareness?.setLocalStateField("button", payload.button);
  };

  destroy() {
    for (const s of this.subscriptions) {
      s();
    }
  }
}

function getVersion(elems: readonly { version: number }[]): number {
  return elems.reduce((acc, curr) => {
    return curr.version + acc;
  }, 0);
}

export class ExcalidrawAssetsBinding {
  subscriptions: (() => void)[] = [];

  constructor(ymap: Y.Map<unknown>, api: ExcalidrawImperativeAPI) {
    this.subscriptions.push(
      api.onChange((_element, _appstate, files) => {
        const doc = ymap.doc;
        doc?.transact(() => {
          for (const key in files) {
            if (!ymap.get(key)) {
              ymap.set(key, files[key]);
            }
          }
        }, this);
      }),
    );

    const handler = (events: Y.YMapEvent<unknown>, txn: Y.Transaction) => {
      if (txn.origin === this) {
        return;
      }

      const addedFiles = [...events.keysChanged].map(
        (key) => ymap.get(key) as BinaryFileData,
      );
      api.addFiles(addedFiles);
    };
    ymap.observe(handler);
    this.subscriptions.push(() => {
      ymap.unobserve(handler);
    });

    api.addFiles([...ymap.keys()].map((key) => ymap.get(key) as BinaryFileData));
  }
  destroy() {
    for (const s of this.subscriptions) {
      s();
    }
  }
}
