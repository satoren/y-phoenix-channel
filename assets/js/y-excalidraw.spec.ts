import { describe, expect, test } from "vitest";
import { syncElementsToYArray } from "./y-excalidraw";
import * as Y from "yjs";

describe("syncElementsToYArray", () => {
  const ydoc = new Y.Doc();
  test("adds 1 + 2 to equal 3", () => {
    const array = ydoc.getArray<Y.Map<unknown>>("elements");
    const e1 = [
      {
        id: "cIQh4r6HOcWQTbIKlVdc7",
        type: "freedraw",
        x: 927.3333740234375,
        y: 171.66668701171875,
        width: 0,
        height: 0,
        angle: 0,
        strokeColor: "#1e1e1e",
        backgroundColor: "transparent",
        fillStyle: "solid",
        strokeWidth: 2,
        strokeStyle: "solid",
        roughness: 1,
        opacity: 100,
        groupIds: [],
        frameId: null,
        roundness: null,
        seed: 1238615798,
        version: 2,
        versionNonce: 1130941290,
        isDeleted: false,
        boundElements: null,
        updated: 1726214549611,
        link: null,
        locked: false,
        points: [[0, 0]],
        pressures: [],
        simulatePressure: true,
        lastCommittedPoint: null,
      },
    ] as const;
    const e2 = [
      {
        id: "cIQh4r6HOcWQTbIKlVdc7",
        type: "freedraw",
        x: 927.3333740234375,
        y: 171.66668701171875,
        width: 76.6666259765625,
        height: 162,
        angle: 0,
        strokeColor: "#1e1e1e",
        backgroundColor: "transparent",
        fillStyle: "solid",
        strokeWidth: 2,
        strokeStyle: "solid",
        roughness: 1,
        opacity: 100,
        groupIds: [],
        frameId: null,
        roundness: null,
        seed: 1238615798,
        version: 22,
        versionNonce: 246133354,
        isDeleted: false,
        boundElements: null,
        updated: 1726214550044,
        link: null,
        locked: false,
        points: [
          [0, 0],
          [-62.66668701171875, 62.666656494140625],
          [-68.66668701171875, 75.33331298828125],
          [-72, 96],
          [-72, 100],
          [-70.66668701171875, 112],
          [-69.3333740234375, 119.33331298828125],
          [-66, 124],
          [-60, 134.66665649414062],
          [-50, 144],
          [-42.66668701171875, 149.33331298828125],
          [-36, 152.66665649414062],
          [-30, 154.66665649414062],
          [-28.66668701171875, 155.33331298828125],
          [-18.66668701171875, 158.66665649414062],
          [-17.3333740234375, 158.66665649414062],
          [-12, 160.66665649414062],
          [-8, 161.33331298828125],
          [-6.66668701171875, 162],
          [3.33331298828125, 162],
          [4.6666259765625, 162],
        ],
        pressures: [],
        simulatePressure: true,
        lastCommittedPoint: null,
      },
    ] as const;
    const e3 = [
      {
        id: "cIQh4r6HOcWQTbIKlVdc7",
        type: "freedraw",
        x: 927.3333740234375,
        y: 171.66668701171875,
        width: 77.33331298828125,
        height: 162.66665649414062,
        angle: 0,
        strokeColor: "#1e1e1e",
        backgroundColor: "transparent",
        fillStyle: "solid",
        strokeWidth: 2,
        strokeStyle: "solid",
        roughness: 1,
        opacity: 100,
        groupIds: [],
        frameId: null,
        roundness: null,
        seed: 1238615798,
        version: 24,
        versionNonce: 1958584298,
        isDeleted: false,
        boundElements: null,
        updated: 1726214550375,
        link: null,
        locked: false,
        points: [
          [0, 0],
          [-62.66668701171875, 62.666656494140625],
          [-68.66668701171875, 75.33331298828125],
          [-72, 96],
          [-72, 100],
          [-70.66668701171875, 112],
          [-69.3333740234375, 119.33331298828125],
          [-66, 124],
          [-60, 134.66665649414062],
          [-50, 144],
          [-42.66668701171875, 149.33331298828125],
          [-36, 152.66665649414062],
          [-30, 154.66665649414062],
          [-28.66668701171875, 155.33331298828125],
          [-18.66668701171875, 158.66665649414062],
          [-17.3333740234375, 158.66665649414062],
          [-12, 160.66665649414062],
          [-8, 161.33331298828125],
          [-6.66668701171875, 162],
          [3.33331298828125, 162],
          [4.6666259765625, 162],
          [5.33331298828125, 162.66665649414062],
          [5.33331298828125, 162.66665649414062],
        ],
        pressures: [],
        simulatePressure: true,
        lastCommittedPoint: [5.33331298828125, 162.66665649414062],
      },
      {
        id: "_FY_CXdx9ffRERbgVngnV",
        type: "freedraw",
        x: 994.6666870117188,
        y: 203.66668701171875,
        width: 94.66668701171875,
        height: 13.33331298828125,
        angle: 0,
        strokeColor: "#1e1e1e",
        backgroundColor: "transparent",
        fillStyle: "solid",
        strokeWidth: 2,
        strokeStyle: "solid",
        roughness: 1,
        opacity: 100,
        groupIds: [],
        frameId: null,
        roundness: null,
        seed: 1837712630,
        version: 13,
        versionNonce: 1690477482,
        isDeleted: false,
        boundElements: null,
        updated: 1726214588511,
        link: null,
        locked: false,
        points: [
          [0, 0],
          [-12, 3.33331298828125],
          [-16.66668701171875, 4.666656494140625],
          [-21.33331298828125, 6],
          [-28, 7.33331298828125],
          [-56.66668701171875, 11.33331298828125],
          [-60, 12],
          [-68.66668701171875, 12.666656494140625],
          [-84, 12.666656494140625],
          [-86.66668701171875, 12.666656494140625],
          [-92.66668701171875, 12.666656494140625],
          [-94.66668701171875, 13.33331298828125],
        ],
        pressures: [],
        simulatePressure: true,
        lastCommittedPoint: null,
      },
    ] as const;
    const e4 = [
      {
        id: "cIQh4r6HOcWQTbIKlVdc7",
        type: "freedraw",
        x: 927.3333740234375,
        y: 171.66668701171875,
        width: 77.33331298828125,
        height: 162.66665649414062,
        angle: 0,
        strokeColor: "#1e1e1e",
        backgroundColor: "transparent",
        fillStyle: "solid",
        strokeWidth: 2,
        strokeStyle: "solid",
        roughness: 1,
        opacity: 100,
        groupIds: [],
        frameId: null,
        roundness: null,
        seed: 1238615798,
        version: 24,
        versionNonce: 1958584298,
        isDeleted: false,
        boundElements: null,
        updated: 1726214550375,
        link: null,
        locked: false,
        points: [
          [0, 0],
          [-62.66668701171875, 62.666656494140625],
          [-68.66668701171875, 75.33331298828125],
          [-72, 96],
          [-72, 100],
          [-70.66668701171875, 112],
          [-69.3333740234375, 119.33331298828125],
          [-66, 124],
          [-60, 134.66665649414062],
          [-50, 144],
          [-42.66668701171875, 149.33331298828125],
          [-36, 152.66665649414062],
          [-30, 154.66665649414062],
          [-28.66668701171875, 155.33331298828125],
          [-18.66668701171875, 158.66665649414062],
          [-17.3333740234375, 158.66665649414062],
          [-12, 160.66665649414062],
          [-8, 161.33331298828125],
          [-6.66668701171875, 162],
          [3.33331298828125, 162],
          [4.6666259765625, 162],
          [5.33331298828125, 162.66665649414062],
          [5.33331298828125, 162.66665649414062],
        ],
        pressures: [],
        simulatePressure: true,
        lastCommittedPoint: [5.33331298828125, 162.66665649414062],
      },
      {
        id: "_FY_CXdx9ffRERbgVngnV",
        type: "freedraw",
        x: 994.6666870117188,
        y: 203.66668701171875,
        width: 104,
        height: 13.33331298828125,
        angle: 0,
        strokeColor: "#1e1e1e",
        backgroundColor: "transparent",
        fillStyle: "solid",
        strokeWidth: 2,
        strokeStyle: "solid",
        roughness: 1,
        opacity: 100,
        groupIds: [],
        frameId: null,
        roundness: null,
        seed: 1837712630,
        version: 20,
        versionNonce: 884557546,
        isDeleted: false,
        boundElements: null,
        updated: 1726214588674,
        link: null,
        locked: false,
        points: [
          [0, 0],
          [-12, 3.33331298828125],
          [-16.66668701171875, 4.666656494140625],
          [-21.33331298828125, 6],
          [-28, 7.33331298828125],
          [-56.66668701171875, 11.33331298828125],
          [-60, 12],
          [-68.66668701171875, 12.666656494140625],
          [-84, 12.666656494140625],
          [-86.66668701171875, 12.666656494140625],
          [-92.66668701171875, 12.666656494140625],
          [-94.66668701171875, 13.33331298828125],
          [-95.33331298828125, 13.33331298828125],
          [-98.66668701171875, 13.33331298828125],
          [-99.33331298828125, 13.33331298828125],
          [-100, 13.33331298828125],
          [-102.66668701171875, 13.33331298828125],
          [-104, 13.33331298828125],
          [-104, 13.33331298828125],
        ],
        pressures: [],
        simulatePressure: true,
        lastCommittedPoint: [-104, 13.33331298828125],
      },
    ] as const;
    syncElementsToYArray(e1, array);
    syncElementsToYArray(e2, array);
    syncElementsToYArray(e3, array);
    syncElementsToYArray(e4, array);

    expect(array.toJSON()).toMatchSnapshot();
  });
});
