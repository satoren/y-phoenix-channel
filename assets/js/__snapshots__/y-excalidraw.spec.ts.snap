// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`syncElementsToYArray > adds 1 + 2 to equal 3 1`] = `
[
  {
    "angle": 0,
    "backgroundColor": "transparent",
    "boundElements": null,
    "fillStyle": "solid",
    "frameId": null,
    "groupIds": [],
    "height": 162.66665649414062,
    "id": "cIQh4r6HOcWQTbIKlVdc7",
    "index": 0,
    "isDeleted": false,
    "lastCommittedPoint": [
      5.33331298828125,
      162.66665649414062,
    ],
    "link": null,
    "locked": false,
    "opacity": 100,
    "points": [
      [
        0,
        0,
      ],
      [
        -62.66668701171875,
        62.666656494140625,
      ],
      [
        -68.66668701171875,
        75.33331298828125,
      ],
      [
        -72,
        96,
      ],
      [
        -72,
        100,
      ],
      [
        -70.66668701171875,
        112,
      ],
      [
        -69.3333740234375,
        119.33331298828125,
      ],
      [
        -66,
        124,
      ],
      [
        -60,
        134.66665649414062,
      ],
      [
        -50,
        144,
      ],
      [
        -42.66668701171875,
        149.33331298828125,
      ],
      [
        -36,
        152.66665649414062,
      ],
      [
        -30,
        154.66665649414062,
      ],
      [
        -28.66668701171875,
        155.33331298828125,
      ],
      [
        -18.66668701171875,
        158.66665649414062,
      ],
      [
        -17.3333740234375,
        158.66665649414062,
      ],
      [
        -12,
        160.66665649414062,
      ],
      [
        -8,
        161.33331298828125,
      ],
      [
        -6.66668701171875,
        162,
      ],
      [
        3.33331298828125,
        162,
      ],
      [
        4.6666259765625,
        162,
      ],
      [
        5.33331298828125,
        162.66665649414062,
      ],
      [
        5.33331298828125,
        162.66665649414062,
      ],
    ],
    "pressures": [],
    "roughness": 1,
    "roundness": null,
    "seed": 1238615798,
    "simulatePressure": true,
    "strokeColor": "#1e1e1e",
    "strokeStyle": "solid",
    "strokeWidth": 2,
    "type": "freedraw",
    "updated": 1726214550375,
    "version": 24,
    "versionNonce": 1958584298,
    "width": 77.33331298828125,
    "x": 927.3333740234375,
    "y": 171.66668701171875,
  },
  {
    "angle": 0,
    "backgroundColor": "transparent",
    "boundElements": null,
    "fillStyle": "solid",
    "frameId": null,
    "groupIds": [],
    "height": 13.33331298828125,
    "id": "_FY_CXdx9ffRERbgVngnV",
    "index": 1,
    "isDeleted": false,
    "lastCommittedPoint": [
      -104,
      13.33331298828125,
    ],
    "link": null,
    "locked": false,
    "opacity": 100,
    "points": [
      [
        0,
        0,
      ],
      [
        -12,
        3.33331298828125,
      ],
      [
        -16.66668701171875,
        4.666656494140625,
      ],
      [
        -21.33331298828125,
        6,
      ],
      [
        -28,
        7.33331298828125,
      ],
      [
        -56.66668701171875,
        11.33331298828125,
      ],
      [
        -60,
        12,
      ],
      [
        -68.66668701171875,
        12.666656494140625,
      ],
      [
        -84,
        12.666656494140625,
      ],
      [
        -86.66668701171875,
        12.666656494140625,
      ],
      [
        -92.66668701171875,
        12.666656494140625,
      ],
      [
        -94.66668701171875,
        13.33331298828125,
      ],
      [
        -95.33331298828125,
        13.33331298828125,
      ],
      [
        -98.66668701171875,
        13.33331298828125,
      ],
      [
        -99.33331298828125,
        13.33331298828125,
      ],
      [
        -100,
        13.33331298828125,
      ],
      [
        -102.66668701171875,
        13.33331298828125,
      ],
      [
        -104,
        13.33331298828125,
      ],
      [
        -104,
        13.33331298828125,
      ],
    ],
    "pressures": [],
    "roughness": 1,
    "roundness": null,
    "seed": 1837712630,
    "simulatePressure": true,
    "strokeColor": "#1e1e1e",
    "strokeStyle": "solid",
    "strokeWidth": 2,
    "type": "freedraw",
    "updated": 1726214588674,
    "version": 20,
    "versionNonce": 884557546,
    "width": 104,
    "x": 994.6666870117188,
    "y": 203.66668701171875,
  },
]
`;
